compatibility/sandbox-migration/runner/Migration/ProposeAccept.hs

```diff
@@ -104,46 +105,26 @@ data ProposeDeal = ProposeDeal
 
 instance A.FromJSON ProposeDeal
 
-data Event
-  = CreatedDeal (ContractId Deal) Deal
-  | ArchivedDeal (ContractId Deal)
-  | CreatedProposeDeal (ContractId ProposeDeal) ProposeDeal
-  | ArchivedProposeDeal (ContractId ProposeDeal)
-  deriving (Eq, Show)
-
-instance A.FromJSON Event where
-    parseJSON = A.withObject "Event" $ \o -> do
-        ty <- o A..: "type"
-        moduleName <- o A..: "moduleName"
-        entityName <- o A..: "entityName"
-        case moduleName of
-            "ProposeAccept" -> case ty of
-                "created" -> case entityName of
-                    "Deal" -> CreatedDeal <$> o A..: "contractId" <*> o A..: "argument"
-                    "ProposeDeal" -> CreatedProposeDeal <$> o A..: "contractId" <*> o A..: "argument"
-                    _ -> fail ("Invalid entity: " <> entityName)
-                "archived" -> case entityName of
-                    "Deal" -> ArchivedDeal <$> o A..: "contractId"
-                    "ProposeDeal" -> ArchivedProposeDeal <$> o A..: "contractId"
-                    _ -> fail ("Invalid entity: " <> entityName)
-                _ -> fail ("Invalid event type: " <> ty)
-            _ -> fail ("Invalid module: " <> moduleName)
-
-data Transaction = Transaction
-  { transactionId :: T.Text
-  , events :: [Event]
-  } deriving (Generic, Eq, Show)
-
-instance A.FromJSON Transaction
-
 data Result = Result
-  { oldProposeDeals :: [Tuple2 (ContractId ProposeDeal) ProposeDeal]
-  , newProposeDeals :: [Tuple2 (ContractId ProposeDeal) ProposeDeal]
-  , oldDeals :: [Tuple2 (ContractId Deal) Deal]
-  , newDeals :: [Tuple2 (ContractId Deal) Deal]
+  { oldProposeDeals :: [Tuple2 ContractId ProposeDeal]
+  , newProposeDeals :: [Tuple2 ContractId ProposeDeal]
+  , oldDeals :: [Tuple2 ContractId Deal]
+  , newDeals :: [Tuple2 ContractId Deal]
   , oldTransactions :: [Transaction]
   , newTransactions :: [Transaction]
   } deriving Generic
 
 instance A.FromJSON Result
 
+
+pattern CreatedProposeDeal :: ContractId -> ProposeDeal -> Event
```

Comment:
Is `pattern` a GHC-specific feature?

Reply:
GHC-specific as in not supported in DAML? No! You can use pattern synonyms in DAML as well (that’s how `Just` works as a synonym for `Some`). However, it can potentially slow your code down over inlining it since our optimizer is shit. For GHC, the optimizer will usually optimize it to the same code.

Reply:
> GHC-specific as in not supported in DAML?

No, as in, part of the extension GHC has on top of standard Haskell.

Just asking, I never saw this. :slightly_smiling_face:

Reply:
Ah yes it is. I don’t differentiate between GHC and standard Haskell. Given that GHC is the only usable Haskell implementation the standard is pretty much irrelevant and it matters much more what GHC does.